target/linux/ath79/files/drivers/mtd/nand/rb4xx_nand.c

```diff
@@ -0,0 +1,269 @@
+		oobregion->offset = 13;
+		oobregion->length = 3;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static int rb4xx_ooblayout_free(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 0;
+		oobregion->length = 4;
+		return 0;
+	case 1:
+		oobregion->offset = 4;
+		oobregion->length = 1;
+		return 0;
+	case 2:
+		oobregion->offset = 6;
+		oobregion->length = 2;
+		return 0;
+	case 3:
+		oobregion->offset = 11;
+		oobregion->length = 2;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static const struct mtd_ooblayout_ops rb4xx_nand_ecclayout_ops = {
+	.ecc = rb4xx_ooblayout_ecc,
+	.free = rb4xx_ooblayout_free,
+};
+
+static int rb4xx_nand_dev_ready(struct mtd_info *mtd)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	return gpiod_get_value_cansleep(info->gpio_rdy);
+}
+
+static void rb4xx_nand_cmd_ctrl(struct mtd_info *mtd, int cmd,
+				unsigned int ctrl)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	u8 data = cmd;
+	int ret;
+
+	if (ctrl & NAND_CTRL_CHANGE) {
+		gpiod_set_value_cansleep(info->gpio_cle, !!(ctrl & NAND_CLE));
+		gpiod_set_value_cansleep(info->gpio_ale, !!(ctrl & NAND_ALE));
+		gpiod_set_value_cansleep(info->gpio_nce,  !(ctrl & NAND_NCE));
+	}
+
+	if (cmd != NAND_CMD_NONE) {
+		ret = rb4xx_cpld_write_nand(info->spi_dev, &data, 1);
+		if (unlikely(ret))
+			pr_err("rb4xx_nand: write cmd failed, error %d\n", ret);
+	}
+}
+
+static unsigned char rb4xx_nand_read_byte(struct mtd_info *mtd)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	unsigned char data;
+	int ret;
+
+	ret = rb4xx_cpld_read_nand(info->spi_dev, &data, 1);
+	if (unlikely(ret)) {
+		pr_err("rb4xx_nand: read data failed, error %d\n", ret);
+		return 0xff;
+	}
+
+	return data;
+}
+
+static void rb4xx_nand_write_buf(struct mtd_info *mtd, const unsigned char *buf,
+				int len)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	int ret;
+
+	ret = rb4xx_cpld_write_nand(info->spi_dev, buf, len);
+	if (unlikely(ret))
+		pr_err("rb4xx_nand: write buf failed, error %d\n", ret);
+}
+
+static void rb4xx_nand_read_buf(struct mtd_info *mtd, unsigned char *buf,
+				int len)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	int ret;
+
+	ret = rb4xx_cpld_read_nand(info->spi_dev, buf, len);
+	if (unlikely(ret))
+		pr_err("rb4xx_nand: read buf failed, error %d\n", ret);
+}
+
+static int rb4xx_nand_probe(struct platform_device *pdev)
+{
+	struct rb4xx_nand_info *info;
+	struct spi_device *spi_dev;
+	struct mtd_info *mtd;
+	int ret;
+
+	info = devm_kzalloc(&pdev->dev, sizeof(*info), GFP_KERNEL);
+	if (!info)
+		return -ENOMEM;
+
+	spi_dev = to_spi_device(pdev->dev.parent);
+	if (!spi_dev) {
+		dev_err(&pdev->dev, "unable to get parent CPLD SPI device\n");
+		return -ENODEV;
+	}
+	info->spi_dev = spi_dev;
```

Comment:
This is not the job of this driver. The CPLD driver must handle its own dependencies.

Reply:
That would mean the cpld driver would need to keep a static pointer to itself, so it knows the spi device to use for these extern function calls? (as we don't have a pointer the private data).

Reply:
BTW, I did [hack up a single driver prototype last night](https://github.com/ch6574/openwrt/commit/1c8478a9acf15df45dc0ca8657ca7b13dd5c46e9), however it panics the kernel as it's unable to find the nand device - I have a  hunch the gpios are not working correctly from earlier in the probe() function?

Reply:
I think you're quite missing the philosophy behind MFD ;)
Here are a few hopefully useful pointers:
- https://bootlin.com/pub/conferences/2015/elce/belloni-mfd-regmap-syscon/belloni-mfd-regmap-syscon.pdf
- https://wiki.analog.com/resources/tools-software/linux-drivers/multifunction-device/adp5520
- https://processors.wiki.ti.com/index.php/TPS65086x_Linux_Driver
- http://www.ti.com/tool/TPS65217SW-LINUX

The general idea is that the mfd bits provide the backbone to interconnect each specific driver. This way you don't need to handle a cascade of pointers to other subsystems said driver shouldn't touch.

HTH

Reply:
Thanks, I'll have a crack at that over the next few days.

Reply:
Curious, is [this](https://github.com/ch6574/openwrt/commit/586d754b60d7d9393c791a4767a98c844e04e39f) the kind of thing you had in mind?

I've hacked up the cpld driver as a mfd, and put the gpio extender underneath it. I can see it works by toggling the connected leds on/off via /sys/class/leds/, so figure I'm doing something right!

Reply:
Yes that's the spirit I think. Beware that the mfd semantics are different from other drivers: I think you'll want to define two cells, one for gpio and one for nand and use the mfd API for the glue between them. But otherwise I think that's the general direction this code needs to take if we want to make it upstreamable.

Disclaimer: as I said I'm not familiar with these subsystems so I hope I'm not talking out of my rear end, however I just stumbled upon this: https://lore.kernel.org/patchwork/patch/553417/

The comments there make me feel more comfortable about suggesting to move toward mfd, although it probably wouldn't hurt to get a second opinion on this :)

Reply:
@f00b4r0 would appreciate if you could take a look at [the other branch](https://github.com/ch6574/openwrt/commits/rb493g-mfd) I have - as that is now a fully working "mfd" setup that I think is a much better solution.

Reply:
N.B. I put your name in as a reviewer as I've been incorporating your feedback from this PR too. Let me know if you want it taken out.

Reply:
Hi, I took a very brief look: it looks a lot better indeed, I like it very much :)

Before I do a deeper review, a couple quick very preliminary remarks:
- try to be more consistent with the typing of your variables and parameters, and above all avoid using bare "unsigned". Likewise, I wouldn't mix "u8" and "unsigned char" but use one or the other consistently throughout the code (my preference goes to "uN" notation for anything that goes to hardware, I think it's clearer). e.g. `rb4xx_cpld_write_cfg()` expects a "u8", don't pass it an "unsigned".
- looking at `rb4xx_gpio_change_cfg()` I think that function looks odd: it's unclear to me why only LED5 needs to be special cased, and looking at the interaction between `rb4xx_gpio_direction_output()`, `rb4xx_gpio_set()`and that function, it seems that the previous two do exactly the same thing (which suggests something is wrong). If only output is supported and that's the static hardware configuration, then the direction_output() should simply be a no-op and return 0. No need to waste time :) You should also consider my previous comment on the way to handle LED5 (assuming the special case is needed), and the fact that the logic should probably be reversed and LED5 be declared as active_low, like the other CPLD LEDs.
- in the header file I think at this time the #defines are unnecessary and should be folded in the respective C files where they are only used
- in `rb4xx_nand_probe()` I think you should not use the index version of devm_gpiod_get() since you name the gpios in the DTS (see how it's done in `drivers/mtd/nand/raw/gpio.c`)? 

Anyway, I think this looks a heck of a lot better than it did, I'd suggest using that branch in the PR and I'll try to do a more in-depth code review soon.

Thanks a lot for your efforts!

Reply:
Thanks again for the feedback, I have just pushed several changes based on this advice.

- Consistent typing. On this I have tried to keep with the same style the subsytem the driver lives in, so the spi and gpio code uses `char`, `unsigned` whereas nand uses `u8` etc. I could be consistent across all 3 drivers, however that would then introduce implicit type casting, which has its own issues. Perhaps I'm thinking about this wrong?
- GPIO code I've re-worked, plus made them all active low (which makes more sense). I'm much more happy with this code now. N.B. `direction_output` and `set` both must be defined according to the docs, as the former is called at setup (and allows an error to be returned), the latter is for all subsequent use (if no setup error).
- Header file is now clean of all defines. Again, happy.
- Lastly the use of indexed gpio in `rb4xx_nand_probe` is interesting. I changed the driver to get by name, however the probe failed saying they didn't exist. I wonder if this driver needs to defer probing until the names are set?
- One other thing I'm facing, is that kernel 4.20 changed the nand api, and so I'm looking at a load of `#ifdef `entries to support two kernels... :frowning_face: 

So, if you are happy to use this code, I'll rebase this PR sometime this week.

Reply:
Hi,
- I found only 5 or 6 occurrences of a lone `unsigned` in `include/linux/spi/spi/h`, I suspect these may be coming from old code. `unsigned` really means `unsigned int` so when you pass an `u8` to such a function (or worse: an `unsigned` to a function that expects an `u8`) you're already doing implicit casting, and that's exactly why the vague `unsigned` should be avoided at all costs IMHO.

- `direction_output()`:
On recent kernels `include/linux/gpio/driver.h` says:
```
 * @get_direction: returns direction for signal "offset", 0=out, 1=in,
 *	(same as GPIOF_DIR_XXX), or negative error.
 *	It is recommended to always implement this function, even on
 *	input-only or output-only gpio chips.
 * @direction_input: configures signal "offset" as input, or returns error
 *	This can be omitted on input-only or output-only gpio chips.
 * @direction_output: configures signal "offset" as output, or returns error
 *	This can be omitted on input-only or output-only gpio chips.
```
Which is why I initially thought it could be omitted. Looking at 4.19, it is indeed necessary, I stand corrected :)
However to be future proof you should probably implement a trivial `get_direction()`.

- gpio probing: it's possible driver ordering is playing tricks, I propose you leave it for now with index calls with maybe a TODO note in the code to fix it? (it has to be clear somewhere that the names in DTS are ignored). We can always deal with that later.

- kernel API changes: they happen. Push your code, we'll see how to limit damage ;)

I'll do a proper code review when the code is PR'd.

Thanks!

Reply:
Thanks, on the `unsigned` point, one example is [spi_transfer](https://github.com/torvalds/linux/blob/master/include/linux/spi/spi.h#L912), where the buffer length `len` is defined as `unsigned` and is something I need to supply - hence  `rb4xx_cpld_write_nand` has that as a parameter.

The calling code in [nand](https://github.com/torvalds/linux/blob/master/include/linux/mtd/rawnand.h#L996) provides a signed `int`, so somewhere in my code I need to convert - either explicitly with a cast, or implictly in the function signatures.

Ditto for GPIO where I have a `size_t` from using ARRAY_SIZE to calculate the buffer size.

On this, I'm unclear what's the preferred way?

(I don't believe there's anywhere that passes `int` / `unsigned` to a `u8`)

Reply:
If you switch to [git blame](https://github.com/torvalds/linux/blame/9cb1fd0efd195590b828b9b865421ad345a4a145/include/linux/spi/spi.h#L912) you'll quickly see that as I mentioned before, this is very old code. No need to forward that old disease in new code. My advice is to explicitly spell out these `unsigned` as `unsigned int`, at the very least. (That particular code clearly calls for `size_t`, but well). Four extra characters can make some problems a lot more obvious. Using a signedness specifier without an explicit size qualifier really is bad style IMO. Regarding the issue of the nand code, since I doubt changing the signatures is a good idea, if I were pedantic I'd add a test for (unlikely()) negative length and return an error if negative, then properly cast to unsigned int before calling the other subsystem.

But please do open a proper PR (or replace the code in this one with the new code): it's becoming tedious to comment that code here and we might lose review history (which is why I'd personally would be in favor of force pushing here, btw) :)

Reply:
Thanks - agree with the proper PR needed as this is a drastic change. Do you prefer a brand new PR so we start with a clean history, or a rebase onto this?

@adrianschmutzler thoughts too?

Reply:
As I said, given there's already a lot of review history here, I'd prefer a rebase. Otherwise this'll be lost. My 2c.

Reply:
I'll rebase it later tonight, and if it looks messy we can always scrap the PR and make a new one at that time.

Reply:
Since @f00b4r0 has done most of the reviewing so far, I leave the decision to him (which he already made).
(I personally would have also accepted a new PR for this case, as there is a lot of progress here and it looks like a lot of stuff will be replaced by the next push.)
Again, thanks for your commitment.